### ISSUE TITLE
[BUG] `not None` semantics is not supported in Pure Python mode for `object` type

### ISSUE THREAD

**Issue by matusvalo** [cython/cython#5624]
### Describe the bug

Compiling reproducer is run without error. This is inconsistent logic used in Python Objects [1].

[1] https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#extension-types-and-none

### Code to reproduce the behaviour:

```python
import typing
import cython

def bar(i: typing.Optional[object]):
    pass

def bar2(i: object):
    pass

bar(None)
bar([1])
bar2(None)    # This line is executed without error
```

### Expected behaviour

Line `bar2(None)` should raise following exception:
```
TypeError: Argument 'i' must not be None
```

### OS

any

### Python version

3.9

### Cython version

master

### Additional context

Cython version:
```
import cython

def bar(object i):
    pass

def bar2(object i not None):
    pass

bar(None)
bar([1])
bar2(None)   # Here is raised TypeError
```

Related to #5612

**Comment by da-woods** [comment]
I don't agree with this (or #5625).

From Cython's point of view, the value of `not None` is to be able to make optimizations by directly accessing an extension type/built-in type/memoryview without having to check if it's `None`. There's really no optimizations to be done for `object` so there's no value in this. The point isn't really to enforce user's type annotations - it's just to make efficient use of them.

I'd also argue that `None` is an `object`, so should be accepted for an `object`/untyped argument.

I'm not sure if non-pure mode accepts `object not None`. I guess it probably does. But that's probably just an accident more than a useful design.

The documentation looks to be in terms of "extension or Python builtin types". I don't really consider `object` to be one of those, so I don't think it's inconsistent. But we could clarify it.

**Comment by matusvalo** [comment]
> I don't agree with this (or https://github.com/cython/cython/issues/5625).

This is fine. I am not 100% sure whether those are real issues (see description). I created the issues because it make sense from consistency stand point - where `not None` is supported in Cython, I expected that also pure python mode has similar behavior.

Feel free to close the issues without asking.

> I'm not sure if non-pure mode accepts object not None.

Cython languages accepts `not None` in both issues as presented in examples (unless I did a stupid mistake)
